### PR TITLE
Fix #477: Config changes so that the backend starts without additional configurations

### LINKED ISSUE #477: Showcase backend won't start with example config
### Describe the bug

A clear and concise description of what the bug is. (provide screenshots if applicable)

### Expected Behavior
Showcase starting with default (example) config 

### What are the steps to reproduce this bug?

```
▶ cp app-config.example.yaml app-config.local.yaml

▶ yarn run start
yarn run v1.22.19
$ turbo run start --parallel
• Packages in scope: app, backend
• Running start in 2 packages
• Remote caching disabled
app:start: cache bypass, force executing 5b0a51096ca854b0
backend:start: cache bypass, force executing 44b2ea957cd8fc87
app:start: $ backstage-cli package start
backend:start: $ backstage-cli package start
app:start: ⚠️   Some of the following packages may be outdated or have duplicate installations:
app:start:
app:start:           @backstage/plugin-catalog-backend, @backstage/plugin-kubernetes, @backstage/plugin-scaffolder-backend
app:start:
app:start: ⚠️   This can be resolved using the following command:
app:start:
app:start:           yarn backstage-cli versions:check --fix
app:start:
backend:start: Build succeeded
app:start: Loaded config from app-config.yaml, app-config.local.yaml
app:start: <i> [webpack-dev-server] Project is running at:
app:start: <i> [webpack-dev-server] Loopback: http://localhost:3000/, http://[::1]:3000/
app:start: <i> [webpack-dev-server] Content not from webpack is served from '/Users/tkral/Code/backstage-showcase/packages/app/public' directory
app:start: <i> [webpack-dev-server] 404s will fallback to '/index.html'
app:start: <i> [webpack-dev-middleware] wait until bundle finished: /
backend:start: Loaded config from app-config.yaml, app-config.local.yaml
backend:start: 2023-08-29T15:38:10.405Z backstage info Found 4 new secrets in config that will be redacted
backend:start: Backend failed to start up Error: Missing required config value at 'backend.auth.keys[0].secret'
backend:start:     at ConfigReader.getString (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/config/dist/index.cjs.js:240:13)
backend:start:     at /Users/tkral/Code/backstage-showcase/node_modules/@backstage/backend-common/dist/index.cjs.js:4034:31
backend:start:     at Array.map (<anonymous>)
backend:start:     at ServerTokenManager.fromConfig (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/backend-common/dist/index.cjs.js:4034:14)
backend:start:     at makeCreateEnv (webpack-internal:///./src/index.ts:66:100)
backend:start:     at main (webpack-internal:///./src/index.ts:118:23)
app:start: webpack compiled successfully
```
adding 
```yaml
backend:
  auth:
    keys: []
```
to `app-config.local.yaml` fixes the issue with `backend.auth.keys[0].secret'` but there is another one 
```
backend:start: Backend failed to start up TypeError: Invalid type in config for key 'proxy.endpoints' in 'app-config.local.yaml', got array, wanted object
backend:start:     at ConfigReader.getOptionalConfig (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/config/dist/index.cjs.js:143:13)
backend:start:     at ConfigReader.getConfig (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/config/dist/index.cjs.js:128:24)
backend:start:     at ObservableConfigProxy.select (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/backend-app-api/dist/index.cjs.js:115:41)
backend:start:     at ObservableConfigProxy.get (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/backend-app-api/dist/index.cjs.js:130:17)
backend:start:     at readProxyConfig (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/plugin-proxy-backend/dist/index.cjs.js:104:94)
backend:start:     at createRouter (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/plugin-proxy-backend/dist/index.cjs.js:136:23)
backend:start:     at async createPlugin (webpack-internal:///./src/plugins/proxy.ts:9:12)
backend:start:     at async addPlugin (webpack-internal:///./src/index.ts:109:191)
backend:start:     at async main (webpack-internal:///./src/index.ts:122:5)
app:start: webpack compiled successfully
```
This can be fixed with 
```yaml
proxy:
  endpoints: {}
```
but than you get

```
backend:start: 2023-08-29T15:41:26.505Z proxy info [HPM] Proxy rewrite rule created: "^/api/proxy/quay/api/?" ~> "/" type=plugin
backend:start: Backend failed to start up Error: Proxy target must be a string
backend:start:     at buildMiddleware (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/plugin-proxy-backend/dist/index.cjs.js:35:11)
backend:start:     at /Users/tkral/Code/backstage-showcase/node_modules/@backstage/plugin-proxy-backend/dist/index.cjs.js:163:9
backend:start:     at Array.forEach (<anonymous>)
backend:start:     at configureMiddlewares (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/plugin-proxy-backend/dist/index.cjs.js:159:31)
backend:start:     at createRouter (/Users/tkral/Code/backstage-showcase/node_modules/@backstage/plugin-proxy-backend/dist/index.cjs.js:137:3)
backend:start:     at async createPlugin (webpack-internal:///./src/plugins/proxy.ts:9:12)
backend:start:     at async addPlugin (webpack-internal:///./src/index.ts:109:191)
backend:start:     at async main (webpack-internal:///./src/index.ts:122:5)
```
This one is not easily fixable in `local` config, the only way I found is to remove `proxy.endpoints` from `app-config.yaml`

### Versions of software used and environment
tested on 7fa8b00d3dc7eb4988307877cef8e9ea1fb32145

### PR DESCRIPTION
**Pull request by jerolimov**
## Description

I got 3 different errors when starting the showcase with just `yarn install` and `yarn start`:

### Error 1: Missing required config value at 'backend.auth.keys[0].secret'

Error:

```log
backend:start: Loaded config from app-config.yaml, app-config.local.yaml
backend:start: 2023-09-04T07:29:18.115Z backstage info Found 4 new secrets in config that will be redacted 
backend:start: Backend failed to start up Error: Missing required config value at 'backend.auth.keys[0].secret'
backend:start:     at ConfigReader.getString (/home/christoph/git/janus-idp/backstage-showcase/node_modules/@backstage/config/dist/index.cjs.js:240:13)
backend:start:     at /home/christoph/git/janus-idp/backstage-showcase/node_modules/@backstage/backend-common/dist/index.cjs.js:4034:31
```

Reason:

This secret in `app-config.yaml` isn't defined by default:

```yaml
backend:
  # Used for enabling authentication, secret is shared by all backend plugins
  # See https://backstage.io/docs/auth/service-to-service-auth for
  # information on the format
  auth:
    keys:
      - secret: ${BACKEND_AUTH_SECRET}
```

Solution / workaround:

I commented out the auth key in `app-config.yaml` and added it enabled to `app-config.produiction.yaml` and `app-config.example.production.yaml` to ensure that it's used for, yes, production installations.

With this change, I got this warning:

```log
backend:start: 2023-09-04T07:33:57.617Z backstage warn Generated a secret for service-to-service authentication: DEVELOPMENT USE ONLY. 
```

But it looks like there is no easy way to enable the NODE_ENV production mode with the CLI. See https://github.com/backstage/backstage/issues/17725

### Error 2: Invalid type in config for key 'proxy.endpoints' in 'app-config.local.yaml', got array, wanted object

```log
backend:start: 2023-09-04T07:33:57.618Z backstage info Created UrlReader predicateMux{readers=azure{host=dev.azure.com,authed=false},bitbucketCloud{host=bitbucket.org,authed=false},bitbucketServer{host=bitbucket.com,authed=false},github{host=github.com,authed=false},gitlab{host=gitlab.com,authed=true},awsS3{host=amazonaws.com,authed=false},fetch{} 
backend:start: Backend failed to start up TypeError: Invalid type in config for key 'proxy.endpoints' in 'app-config.local.yaml', got array, wanted object
backend:start:     at ConfigReader.getOptionalConfig (/home/christoph/git/janus-idp/backstage-showcase/node_modules/@backstage/config/dist/index.cjs.js:143:13)
backend:start:     at ConfigReader.getConfig (/home/christoph/git/janus-idp/backstage-showcase/node_modules/@backstage/config/dist/index.cjs.js:128:24)
```

Solution:

Updated `app-config.example.yaml`

```diff
proxy:
-  endpoints: []
+  endpoints: {}
```

### Error 3: Proxy target for /api/proxy/sonarqube must be a string, but is undefined

The origin error message was a bit shorter, but I made it more verbose in `node_modules/@backstage/plugin-proxy-backend/dist/index.cjs.js`. Raised an upstream PR for this:

* https://github.com/backstage/backstage/pull/19759

```
backend:start: Backend failed to start up Error: Proxy target for /api/proxy/sonarqube must be a string, but is undefined
backend:start:     at buildMiddleware (/home/christoph/git/janus-idp/backstage-showcase/node_modules/@backstage/plugin-proxy-backend/dist/index.cjs.js:35:11)
```

The reason is that some variables in multiple plugins aren't defined without additional configuration work:

```yaml
    # Plugin: SonarQube
    '/sonarqube':
      # if not using sonarqube, assign SONARQUBE_URL to https://sonarcloud.io
      # target: https://sonarcloud.io/api
      target: ${SONARQUBE_URL}/api
      allowedMethods: ['GET']
      auth: ${SONARQUBE_TOKEN}
```

Solution / workaround:

To skip the error, I added this workaround to `app-config.example.yaml`:

```diff
proxy:
+  skipInvalidProxies: true
-  endpoints: []
+  endpoints: {}
```

After copying the example again over my local config, I got this warning:

```log
backend:start: 2023-09-04T07:58:24.621Z proxy warn skipped configuring /sonarqube due to Proxy target for /api/proxy/sonarqube must be a string, but is undefined type=plugin
backend:start: 2023-09-04T07:58:24.621Z proxy warn skipped configuring /jenkins/api due to Proxy target for /api/proxy/jenkins/api must be a string, but is undefined type=plugin
backend:start: 2023-09-04T07:58:24.622Z proxy warn skipped configuring /jira/api due to Proxy target for /api/proxy/jira/api must be a string, but is undefined type=plugin
backend:start: 2023-09-04T07:58:24.622Z proxy warn skipped configuring /jfrog-artifactory/api due to Proxy target for /api/proxy/jfrog-artifactory/api must be a string, but is undefined type=plugin
backend:start: 2023-09-04T07:58:24.622Z proxy info [HPM] Proxy created: /pagerduty  -> https://api.pagerduty.com type=plugin
backend:start: 2023-09-04T07:58:24.622Z proxy info [HPM] Proxy rewrite rule created: "^/api/proxy/pagerduty/?" ~> "/" type=plugin
backend:start: 2023-09-04T07:58:24.623Z proxy warn skipped configuring /dynatrace due to Proxy target for /api/proxy/dynatrace must be a string, but is undefined type=plugin

```

And finally, a working showcase.

This is my first PR. So please let me know if additional updates to some docs are required. And especially if the changes to the production files are fine.

## Which issue(s) does this PR fix

- Fixes #477

## PR acceptance criteria

Please make sure that the following steps are complete:

- GitHub Actions are completed and successful
- Unit Tests are updated and passing
- E2E Tests are updated and passing
- Documentation is updated if necessary
- Add a screenshot if the change is UX/UI related

### CODE REVIEW

File: app-config.yaml [hunk 1]
```diff
@@ -12,12 +12,13 @@ organization:
...
 backend:
-  # Used for enabling authentication, secret is shared by all backend plugins
+  # Required for authentication. This secret is shared by all backend plugins.
+  # In development mode a temporary secret is automatically created.
   # See https://backstage.io/docs/auth/service-to-service-auth for
   # information on the format
-  auth:
-    keys:
-      - secret: ${BACKEND_AUTH_SECRET}
+#  auth:
```
[the remark this answers is not in the document]
Comment (jerolimov) [reply app-config.yaml]: Thanks for the review. I applied that change.

File: app-config.production.yaml [hunk 2]
```diff
@@ -3,6 +3,12 @@ app:
   baseUrl: http://localhost:7007
 
 backend:
+  # Required for authentication. This secret is shared by all backend plugins.
+  # See https://backstage.io/docs/auth/service-to-service-auth for
+  # information on the format
+  auth:
+    keys:
+      - secret: ${BACKEND_AUTH_SECRET}
```
Comment (tumido) [inline app-config.production.yaml:11]: Same as above.

File: app-config.example.production.yaml [hunk 3]
```diff
@@ -3,6 +3,12 @@ app:
   baseUrl: http://localhost:7007
 
 backend:
+  # Required for authentication. This secret is shared by all backend plugins.
+  # See https://backstage.io/docs/auth/service-to-service-auth for
+  # information on the format
+  auth:
+    keys:
+      - secret: ${BACKEND_AUTH_SECRET}
```
[the remark this answers is not in the document]
Comment (kadel) [reply app-config.example.production.yaml:11]: Let's comment it out here.
I'm currently looking at the ways we can better organize configs in the container image and helm chart.

**Review by schultzp2020 — APPROVED** [review]
LGTM

**Review by tumido — APPROVED** [review]
/lgtm
/approve

**Review by christophe-f — APPROVED** [review]

**Comment by josephca** [comment]
@christophe-f @nickboldt shouldn't this fix be backported into 0.2.x branch? I'm experiencing the similar issue with the 0.2 image.

**Comment by christophe-f** [comment]
Yes @josephca. I will back port a couple of commits tomorrow that contains fixes for 0.2

**Comment by jerolimov** [comment]
@christophe-f you will take care of the backport, or should I open a cherry-pick PR? Or do we have a bot for this?

**Pull request merged**